// vitepress-blog@1.1.0-alpha.4/dist/theme-helper/hooks/useFriendlyDate.js
/*! 
 * vitepress-blog v1.1.0-alpha.3
 * (c) yue1123
 * Homepage: https://github.com/yue1123/vitepress-blog
 * Released under the MIT License.
 */
import { computed } from 'vue';

function useFriendlyDate(_date) {
  return computed(() => {
    const date = new Date(_date);
    var formats = {
      year: "%n%\u5E74\u524D",
      month: "%n%\u4E2A\u6708\u524D",
      day: "%n%\u5929\u524D",
      hour: "%n%\u5C0F\u65F6\u524D",
      minute: "%n%\u5206\u949F\u524D",
      second: "%n%\u79D2\u524D"
    };
    var now = Date.now();
    var seconds = Math.floor((now - date.getTime()) / 1e3);
    var minutes = Math.floor(seconds / 60);
    var hours = Math.floor(minutes / 60);
    var days = Math.floor(hours / 24);
    var months = Math.floor(days / 30);
    var diffType = "day";
    var diffValue = 0;
    if (months > 0) {
      return `${date.getFullYear()}-${date.getMonth() + 1}-${date.getDate()}`;
    } else {
      if (days > 0) {
        diffType = "day";
        diffValue = days;
      } else {
        if (hours > 0) {
          diffType = "hour";
          diffValue = hours;
        } else {
          if (minutes > 0) {
            diffType = "minute";
            diffValue = minutes;
          } else {
            diffType = "second";
            diffValue = seconds === 0 ? seconds = 1 : seconds;
          }
        }
      }
    }
    return formats[diffType].replace("%n%", diffValue + "");
  });
}

export { useFriendlyDate };
