// vitepress-blog@1.1.0-alpha.4/dist/main/index.js
/*! 
 * vitepress-blog v1.1.0-alpha.3
 * (c) yue1123
 * Homepage: https://github.com/yue1123/vitepress-blog
 * Released under the MIT License.
 */
import { resolve } from 'path';
import blogHelper from './plugins/blog-helper.js';
import { expandMarkdownIt } from './plugins/expand-markdown-it.js';
import deepmerge from 'deepmerge';

const getSharedConfig = (userConfig) => {
  return {
    cacheDir: resolve(process.cwd(), "./node_modules/.vitepress"),
    markdown: {
      config: (md) => expandMarkdownIt(md)
    },
    vite: { plugins: [blogHelper(userConfig)] },
    themeConfig: {
      outline: [1, 5],
      sortby: "CREATE_TIME",
      titleOrder: "fileName",
      pagination: {
        prevText: "\u4E0A\u4E00\u9875",
        nextText: "\u4E0B\u4E00\u9875",
        size: 10,
        pagerCount: 7,
        hideOnSinglePage: true
      }
    },
    head: [
      [
        "link",
        {
          rel: "stylesheet",
          href: "https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.5.1/katex.min.css"
        }
      ]
    ]
  };
};
function defineConfig(config) {
  return deepmerge(getSharedConfig(config), config);
}

export { defineConfig };
